dali/operators/random/rng_base_gpu.h

```diff
@@ -106,64 +105,58 @@ std::pair<std::vector<int>, int> DistributeBlocksPerSample(
   int64_t block = 0;
   for (int s = 0; s < shape.size(); s++) {
-    T *sample_data = static_cast<T *>(output[s].data);
     auto sample_size = volume(shape[s]);
     if (sample_size == 0)
       continue;
     auto work_per_block = div_ceil(sample_size, blocks_per_sample[s]);
     int64_t offset = 0;
     for (int b = 0; b < blocks_per_sample[s]; ++b, ++block) {
       blocks[block].sample_idx = s;
-      blocks[block].output = sample_data + offset;
-      blocks[block].size = std::min(work_per_block, sample_size - offset);
-      offset += blocks[block].size;
+      blocks[block].p_offset = offset;
+      blocks[block].p_count = std::min(work_per_block, sample_size - offset);
+      offset += blocks[block].p_count;
     }
   }
   return blocks_num;
 }
 
 template <typename T>
-int64_t SetupBlockDescs(BlockDesc<true> *blocks, int64_t block_sz, int64_t max_nblocks,
-                        TensorListView<StorageGPU, T> &output,
-                        TensorListView<StorageGPU, const T> &input) {
-  std::vector<int> blocks_per_sample;
-  int64_t blocks_num;
-  auto &shape = output.shape;
-  assert(output.shape == input.shape);
-  std::tie(blocks_per_sample, blocks_num) = DistributeBlocksPerSample(shape, block_sz, max_nblocks);
-  int64_t block = 0;
-  for (int s = 0; s < shape.size(); s++) {
+int64_t SetupSampleDescs(SampleDesc *samples,
+                         TensorListView<StorageGPU, T> &output,
+                         TensorListView<StorageGPU, const T> &input,
+                         int channel_dim = -1) {
+  int nsamples = output.num_samples();
+  for (int s = 0; s < nsamples; s++) {
     T *sample_out = static_cast<T*>(output[s].data);
     const T *sample_in = static_cast<const T*>(input[s].data);
-    auto sample_size = volume(shape[s]);
-    if (sample_size == 0)
-      continue;
-    auto work_per_block = div_ceil(sample_size, blocks_per_sample[s]);
-    int64_t offset = 0;
-    for (int b = 0; b < blocks_per_sample[s]; ++b, ++block) {
-      blocks[block].sample_idx = s;
-      blocks[block].output = sample_out;
-      blocks[block].input = sample_in;
-      blocks[block].offset = offset;
-      blocks[block].size = std::min(work_per_block, sample_size - offset);
-      offset += blocks[block].size;
+    samples[s].output = sample_out;
+    samples[s].input = sample_in;
+    auto sh = output.shape.tensor_shape_span(s);
+    int64_t sample_sz = volume(sh);
+    if (channel_dim >= 0) {
+      int nchannels = sh[channel_dim];
+      samples[s].p_count = sample_sz / nchannels;
+      samples[s].p_stride = channel_dim == 0 ? 1 : nchannels;
+      samples[s].c_count = nchannels;
+      samples[s].c_stride = volume(sh.begin() + channel_dim + 1, sh.end());
+    } else {
+      samples[s].p_count = sample_sz;
+      samples[s].p_stride = 1;
+      samples[s].c_count = 1;
+      samples[s].c_stride = 1;
     }
   }
-  return blocks_num;
+  return nsamples;
```

Comment:
Does it make sense to return just output.num_samples()?

Reply:
it's exactly that. I just put it in a variable to be used in the loop.

Reply:
ok